### ISSUE TITLE
Proposal: split server from project

### ISSUE THREAD

**Issue by Gazler** [beam-telemetry/telemetry_metrics_prometheus#14]
Thanks for this library. I'd like to preface by saying that I have read https://github.com/bryannaegele/telemetry_metrics_prometheus/blob/master/docs/rationale.md#why-include-a-server

I'd like to propose splitting this library out into 2 parts, one dealing with the conversion of telemetry events to a prometheus friendly format, the other dealing with the web server side of things.

I agree with the inclusion of a server in terms of getting up and running quickly, however I think there are some downsides to this approach that make using the library difficult for certain use cases:

 * Using a different version of cowboy (or gun which depends on cowlib) causing dependency issues.
 * Using a different authentication method such as an IP whitelist or HTTP basic auth.
 * Environments where exposing a different port is not possible, such as Heroku.

This library already does a lot of heavy lifting in handling the telemetry events. I think it could be split in 2 ways.

### Extraction of web:

 1. Remove the Router and server code from `telemetry_metrics_prometheus` into a separate project (such as `telemetry_metrics_prometheus_web` or `telemetry_metrics_prometheus_plug`)
 2. Depend on `telemetry_metrics_prometheus` from `telemetry_metrics_prometheus_web`
 3. Using `telemetry_metrics_prometheus_web` behaves as it does today, running its own webserver.

### Extraction of core:

 1. Leave the Router and server code in `telemetry_metrics_prometheus` and extract everything else into `telemetry_metrics_prometheus_core`
 2. Depend on `telemetry_metrics_prometheus_core` from `telemetry_metrics_prometheus`
 3. Using `telemetry_metrics_prometheus` behaves as it does today, users with different needs can use `telemetry_metrics_prometheus_core`.

If you agree with this proposal then I am happy to help in whatever way I can.

**Comment by bryannaegele** [comment]
Hey @Gazler! Thanks for opening this.

This is something that was on my mind from the beginning and wanted to see how adoption went. I've had a few requests now to split it, so I think it's time to do so!

Of the two options, the extraction of the core would be the least disruptive to existing users and still fit what I've seen as the most common use case thus far while also fitting the goal of reporters to be all-inclusive, so I feel this is the route to go. 👍 

> Using a different version of cowboy (or gun which depends on cowlib) causing dependency issues.

I hope relaxing the dependency back down to "~> 2.0" could alleviate this issue a bit, though I'm not sure about introducing an adapter pattern right away. What are your thoughts? It's still really early, so it might be nice to see if people start adopting core over this package to start and reassess. You have more experience with maintaining `plug_cowboy`, so I'd like to get your thoughts.

> Using a different authentication method such as an IP whitelist or HTTP basic auth.

This is an item on my roadmap but was hoping someone would step up to tackle it since it wasn't an immediate need at work. I'd like to see this package flesh out to pass any available option to `plug_cowboy` (most importantly https support) and potentially passing plugs. When you get down to it, the server setup provided here is quite minimal for a user to implement themselves. Your help in fleshing out this area post-split would be greatly appreciated! ❤️

**Comment by Gazler** [comment]
Awesome, I can make a start on this, I notice you've already created the other repo, I can't open a PR on there because there's nothing there. :( I'll push to my own repo and then you can clone it and change push to your remote if that works for you.

> > Using a different version of cowboy (or gun which depends on cowlib) causing dependency issues.
> 
> I hope relaxing the dependency back down to "~> 2.0" could alleviate this issue a bit, though I'm not sure about introducing an adapter pattern right away. What are your thoughts? It's still really early, so it might be nice to see if people start adopting core over this package to start and reassess. You have more experience with maintaining `plug_cowboy`, so I'd like to get your thoughts.

I don't think it matters too much, for most users, a compatible version of cowboy/cowlib will be found. However for users outside of this case, the inclusion of the server will prevent them using this library. However with the core extraction, it will allow them to use all of the core features.

> > Using a different authentication method such as an IP whitelist or HTTP basic auth.
> 
> This is an item on my roadmap but was hoping someone would step up to tackle it since it wasn't an immediate need at work. I'd like to see this package flesh out to pass any available option to `plug_cowboy` (most importantly https support) and potentially passing plugs. When you get down to it, the server setup provided here is quite minimal for a user to implement themselves. Your help in fleshing out this area post-split would be greatly appreciated!

Same as above here, I was talking about users being able to use their own server for this, however there's no reason these features couldn't be added to this library. Adding things like IP whitelist and HTTP basic auth will either involve bringing in additional dependencies or re-implementing a subset of libraries such as https://hex.pm/packages/cidr. For HTTP Basic auth, the prometheus_plugs library implements it here: https://github.com/deadtrickster/prometheus-plugs/blob/72fa654abd1222ee298e13051035d874fcaa1adf/lib/prometheus/plug_exporter.ex#L194

The way Phoenix handles https with cowboy is defined in this module https://github.com/phoenixframework/phoenix/blob/4bf53ecaae5a9057ea57c248964490dfdee312af/lib/phoenix/endpoint/cowboy2_adapter.ex#L76

**Comment by Gazler** [comment]
I've made the required changes

 * https://github.com/Gazler/telemetry_metrics_prometheus_core/pull/1
 * https://github.com/bryannaegele/telemetry_metrics_prometheus/pull/15

I have done this in a way that git history is preserved for both projects, which should make it easy to perform blame/log/bisect/whatever.

If you are happy with core, you can run the following to push it to your repo (apologies if this is obvious):

 * `git clone git@github.com:Gazler/telemetry_metrics_prometheus_core.git -b feat/remove-server`
 * `cd telemetry_metrics_prometheus_core`
 * `git checkout -b master`
 * `git remote rm origin`
 * `git remote add origin git@github.com:bryannaegele/telemetry_metrics_prometheus_core.git`
 * `git push origin master`

This PR will need to be updated to point to the correct git repo (or hex package if we publish). https://github.com/bryannaegele/telemetry_metrics_prometheus/pull/15

One thing that might be worth changing in the core project (I can open an issue there when it is set up) is changing it to use a `child_spec/1` function so that it is started under the users supervision tree (we can decide how to handle that in the web project too.) Using child_spec will work more closely to how https://github.com/arkgil/telemetry_metrics_statsd and the `Telemetry.Metrics.ConsoleReporter` work.

**Comment by bryannaegele** [comment]
Oh, man! I was working on this last night and was waiting to push the core up until I had both done. 😆 

I'll compare and get core pushed up here shortly, then we can work through the PR here.

**Comment by Gazler** [comment]
Ah, that's unfortunate. I have tested with the core implementation I did
which is at
https://github.com/Gazler/telemetry_metrics_prometheus_core/pull/1

The code is probably the same (I don't have time to diff just now). I used
this project as a base so that commit history will be preserved, that's
probably the only difference though, I don't mind which one we use, I
needed to do the split anyway to get #15 working.
